File: src/05.linkList/07.206.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

import { ListNode } from "./01.203";

export function reverseList(head: ListNode | null): ListNode | null {
    // 迭代
    // if (!head) {
    //     return head;
    // }

    // let curNode = head;
    // let prevNode = null;
    // while (curNode) {
    //     const next = curNode.next;
    //     curNode.next = prevNode;
    //     prevNode = curNode;
    //     curNode = next;
    // }

    // return prevNode;

    // 递归
    if (!head || !head.next) {
        return head;
    }
    
    const newHead = reverseList(head.next);
    // 先递后归, 当找到最后一个节点时, 才会依次执行下面的代码
    head.next.next = head;
    head.next = null;
    return newHead;
}

File: src/05.linkList/07.206.go
package linklist

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func ReverseList(head *ListNode) *ListNode {
	// 递归
	// if head == nil || head.Next == nil {
	// 	return head
	// }
	// newHead := ReverseList(head.Next)
	// head.Next.Next = head
	// head.Next = nil
	// return newHead

	// 迭代
	if head == nil {
		return head
	}

	var prevNode *ListNode
	curNode := head

	for curNode != nil {
		next := curNode.Next
		curNode.Next = prevNode
		prevNode = curNode
		curNode = next
	}

	return prevNode
}
